### PR TITLE
Destination Checkpointing: Add StateMessage handing to `BufferedStreamConsumer`

### PR DESCRIPTION
**Pull request by cgardens**
## What
* `BufferedStreamConsumer` no longer uses an on-disk queue. Relies on STDIN to provide back pressure on the worker as needed.
* `BufferedStreamConsumer` keeps track of the last state message it has received from the worker as well as the last record that was committed to a temporary table.
* This does NOT implement checkpointing yet. It is an intermediate step that is safe to merge before we roll out that feature. I have left stubs pointing out where checkpointing will be added in.

### CODE REVIEW

File: airbyte-commons/src/test/java/io/airbyte/commons/concurrency/ThreadTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+package io.airbyte.commons.concurrency;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+@Disabled
+public class ThreadTest {
```
Comment (michel-tricot) [inline airbyte-commons/src/test/java/io/airbyte/commons/concurrency/ThreadTest.java:33]: What is that file?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 2]
```diff
@@ -24,93 +24,101 @@
...
+                         Consumer<AirbyteStateMessage> checkpointConsumer) {
+    this.checkpointConsumer = checkpointConsumer;
     this.hasStarted = false;
+    this.hasClosed = false;
     this.onStart = onStart;
     this.recordWriter = recordWriter;
     this.onClose = onClose;
     this.catalog = catalog;
-    this.pairs = pairs;
+    this.pairs = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog);
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:118]: This field name is very unclear. A reader won't know what it is or what it is used for.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 3]
```diff
@@ -123,103 +131,98 @@ protected void startTracked() throws Exception {
...
+      if (!isValidRecord.apply(data)) {
+        pairToIgnoredRecordCount.put(stream, pairToIgnoredRecordCount.getOrDefault(stream, 0L) + 1L);
+        return;
+      }
+
+      if (queue.size() == BATCH_SIZE) {
+        flushQueueToDestination();
+      }
+
+      if (!queue.offer(message)) {
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:159]: So we have one global queue now?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 4]
```diff
@@ -24,93 +24,101 @@
...
     this.hasStarted = false;
+    this.hasClosed = false;
     this.onStart = onStart;
     this.recordWriter = recordWriter;
     this.onClose = onClose;
     this.catalog = catalog;
-    this.pairs = pairs;
+    this.pairs = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog);
     this.isValidRecord = isValidRecord;
+    this.queue = new ArrayBlockingQueue<>(BATCH_SIZE);
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:120]: You actually don't need a blocking queue you since your producing and consuming in the same thread and in the same function. Just a regular array list would work here.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java (hunk 3 above)
Comment (michel-tricot) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:159]: What is the tradeoff vs the previous implementation?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 5]
```diff
@@ -123,103 +131,98 @@ protected void startTracked() throws Exception {
...
-    if (!pairs.contains(pair)) {
-      throw new IllegalArgumentException(
-          String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
-              Jsons.serialize(catalog), Jsons.serialize(message)));
+  }
+
+  private void flushQueueToDestination() throws Exception {
+    final List<AirbyteMessage> queueContents = new ArrayList<>();
+    queue.drainTo(queueContents);
+    final Map<Type, List<AirbyteMessage>> recordsByType = queueContents
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:173]: Why do you need that since you only enqueuing RECORDS?

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 6]
```diff
@@ -123,103 +131,98 @@ protected void startTracked() throws Exception {
...
+      if (!pairs.contains(stream)) {
+        throwUnrecognizedStream(catalog, message);
+      }
+
+      if (!isValidRecord.apply(data)) {
+        pairToIgnoredRecordCount.put(stream, pairToIgnoredRecordCount.getOrDefault(stream, 0L) + 1L);
+        return;
+      }
+
+      if (queue.size() == BATCH_SIZE) {
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:155]: It better to flush as soon as possible instead of waiting for the next call to accept. So I would move that test after `offer`

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java [hunk 7]
```diff
@@ -123,103 +131,98 @@ protected void startTracked() throws Exception {
...
+    final Map<Type, List<AirbyteMessage>> recordsByType = queueContents
+        .stream()
+        .collect(Collectors.groupingBy(AirbyteMessage::getType));
+
+    final List<AirbyteRecordMessage> records = recordsByType.getOrDefault(Type.RECORD, new ArrayList<>())
+        .stream()
+        .map(AirbyteMessage::getRecord)
+        .collect(Collectors.toList());
+
+    final Map<AirbyteStreamNameNamespacePair, List<AirbyteRecordMessage>> recordsByStream = records.stream()
```
Comment (michel-tricot) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:182]: Instead of doing all this complex logic, why do just partition your queue?

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalWorkflowTest.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.function.Consumer;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+// based of example in temporal samples:
+// https://github.com/temporalio/samples-java/blob/master/src/test/java/io/temporal/samples/hello/HelloActivityTest.java
+@Disabled
+class TemporalWorkflowTest {
```
Comment (michel-tricot) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalWorkflowTest.java:44]: Why is that file here?

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TestWorkflow.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@WorkflowInterface
+public interface TestWorkflow {
+
+  @WorkflowMethod
+  String run();
+
+  class WorkflowImpl implements TestWorkflow {
```
Comment (michel-tricot) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/TestWorkflow.java:45]: Same

**Review by michel-tricot — CHANGES_REQUESTED** [review]

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TestWorkflow.java (hunk 9 above)
Comment (cgardens) [reply airbyte-workers/src/test/java/io/airbyte/workers/temporal/TestWorkflow.java]: sorry. this doesn't belong in this PR. removed.

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalWorkflowTest.java (hunk 8 above)
Comment (cgardens) [reply airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalWorkflowTest.java]: sorry. this doesn't belong in this PR. removed.

File: airbyte-commons/src/test/java/io/airbyte/commons/concurrency/ThreadTest.java (hunk 1 above)
Comment (cgardens) [reply airbyte-commons/src/test/java/io/airbyte/commons/concurrency/ThreadTest.java]: sorry. this doesn't belong in this PR. removed.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java (hunk 5 above)
Comment (cgardens) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:173]: agreed. a simplification i made from the original impl i was considering makes this unnecessary.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java (hunk 7 above)
Comment (cgardens) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:182]: Are you talking about the record type stuff (that part i 100% agree) or partitioning by stream? In the latter case, I don't think that gets us anything (or really matters). Marginally easier to check the queue size with it not partitioned.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java (hunk 3 above)
Comment (cgardens) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java:159]: Good question. Keep in mind that the switching to a global queue is a tradeoff is informed by the fact that our state objects are global constrains us to think this way. I'm not changing it because I prefer it for any particular technical or performance reason.

The trade off is that previously we could flush data for any stream to a tmp table independently of any other stream. We were allowed to do this, because the destination did not care about state (we were all or nothing). Because state objects are global, we need to treat the data as if it has a global ordering so that when declare a state message "committed", we know we actually added all of its records. In practice, I do not think this will result in a big performance tradeoff. The worker is already affectively a global queue, we are just pushing that single queue concept a little further into the destination.

There are definitely pathological cases that can make this perform poorly. For example if you had 10K streams and you round robin-ed a record from each stream, it would mean each time we flush data to the temp tables we would be flushing one record to 10K individual tables. I think this pathological case will be rare however. Anecdotally most sources group the records they output by stream already. CDC is a notable exception here, but again, I'm not sure, given the global state object, that there's too much we can do about it. 

If this does become an issue, we can move to a more robust but complex solution where we do per stream queues. The trick here is managing queue size smartly. Likely we would either need to do something clever or go back to spilling to disk, which has its own set of drawbacks.

We can change this approach as we introduce stream-level state objects, but sources that only support global state (e.g. CDC) are going to be constrained to this sort of thinking.

**Review by michel-tricot — COMMENTED** [review]
LGTM % the unaddressed comments

**Review by michel-tricot — APPROVED** [review]
